coalib/core/Core.py

```diff
@@ -327,34 +150,177 @@ def instantiate_and_track(prev_bear_type, next_bear_type):
-    executor = concurrent.futures.ProcessPoolExecutor()
-
-    # Initialize dependency tracking.
-    dependency_tracker, bears_to_schedule = initialize_dependencies(bears)
-
-    # Let's go.
-    schedule_bears(bears_to_schedule, result_callback, dependency_tracker,
-                   event_loop, {}, executor)
-    try:
-        event_loop.run_forever()
-    finally:
-        event_loop.close()
+        """
+        self.bears = bears
+        self.result_callback = result_callback
+
+    def run(self):
+        """
+        Runs a coala session.
+        """
+        # FIXME Allow to pass different executors nicely, for example to execute
+        # FIXME   coala with less cores, or to schedule jobs on distributed
+        # FIXME   systems (for example Mesos).
+
+        # Set up event loop and executor.
+        self.event_loop = asyncio.SelectorEventLoop()
+        self.executor = concurrent.futures.ProcessPoolExecutor()
+        self.running_tasks = {}
+
+        # Initialize dependency tracking.
+        self.dependency_tracker, bears_to_schedule = initialize_dependencies(
+            self.bears)
+        # Let's go.
+        self.schedule_bears(bears_to_schedule)
+        try:
+            self.event_loop.run_forever()
+        finally:
+            self.event_loop.close()
+
+    def schedule_bears(self, bears):
+        """
+        Schedules the tasks of bears to the given executor and runs them on the
+        given event loop.
+
+        :param bears:
+            A list of bear instances to be scheduled onto the process pool.
+        """
+        for bear in bears:
+            if self.dependency_tracker.get_dependencies(
+                    bear):  # pragma: no cover
+                logging.warning(
+                    'Dependencies for {!r} not yet resolved, holding back. '
+                    'This should not happen, the dependency tracking system '
+                    'should be smarter. Please report this to the developers.'
+                    .format(bear))
+            else:
+                tasks = {
+                    self.event_loop.run_in_executor(
+                        self.executor, bear.execute_task,
+                        bear_args, bear_kwargs)
+                    for bear_args, bear_kwargs in bear.generate_tasks()}
+
+                self.running_tasks[bear] = tasks
+
+                for task in tasks:
+                    task.add_done_callback(functools.partial(
+                        self.finish_task, bear))
+
+                logging.debug('Scheduled {!r} (tasks: {})'.format(bear,
+                                                                  len(tasks)))
+
+                if not tasks:
+                    # We need to recheck our runtime if something is left to
+                    # process, as when no tasks were offloaded the event-loop
+                    # could hang up otherwise.
+                    self.cleanup_bear(bear)
+
+    def cleanup_bear(self, bear):
```

Comment:
I'm thinking of prepending an underscore to the method names, to mark them as "private".
What do you think?